### PR TITLE
🎲 For #6907 - Dispatches tabsTray update when MediaState changes.

### PR DESCRIPTION
**Pull request by boek**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/tabs/src/test/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenterTest.kt [hunk 1]
```diff
@@ -216,6 +218,33 @@ class TabsTrayPresenterTest {
         verify(tabsTray).onTabsChanged(3, 1)
     }
 
+    @Test
+    fun `tabs tray will get updated if mediaState changes`() {
+        val store = BrowserStore(
+                BrowserState(
```
Comment (gabrielluong) [inline components/feature/tabs/src/test/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenterTest.kt:224]: Possible nit is that we are using 8 spaces vs 4 spaces

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt [hunk 2]
```diff
@@ -31,14 +31,21 @@ class TabsTrayPresenter(
...
+    private var mediaScope: CoroutineScope? = null
+    private var tabScope: CoroutineScope? = null
 
     fun start() {
-        scope = store.flowScoped { flow -> collect(flow) }
+        tabScope = store.flowScoped { flow -> collect(flow) }
+        mediaScope = store.flowScoped {
+            it
+                .ifChanged { state -> state.media.aggregate }
+                .collect { this.tabs?.also(tabsTray::updateTabs) }
```
Comment (jonalmeida) [inline components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:42]: This evades the `tabFilter`, see above comment on how we can fix this and the previous concern.

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt [hunk 3]
```diff
@@ -31,14 +31,21 @@ class TabsTrayPresenter(
...
 ) {
     private var tabs: Tabs? = null
-    private var scope: CoroutineScope? = null
+    private var mediaScope: CoroutineScope? = null
+    private var tabScope: CoroutineScope? = null
 
     fun start() {
-        scope = store.flowScoped { flow -> collect(flow) }
+        tabScope = store.flowScoped { flow -> collect(flow) }
+        mediaScope = store.flowScoped {
```
[the remark this answers is not in the document]
Comment (csadilek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:39]: We looked into `ifAnyChanged` and found two scopes easier to read esp. as one change is for `tabs` the other on the root browser state (`mediastate`). This shouldn't cause duplicate notifications though? What am I missing? :) Do the tabs change if the media state changes?

[the remark this answers is not in the document]
Comment (csadilek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:39]: So, in other words. If one flow only observes `tabs` it should not be affected from changed to the `mediastate`, but I may be missing something.

[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:39]: Thanks for clarifying @csadilek!

@boek I'm assuming that Fenix is retrieving the media state when the `updateTabs` is invoked? It doesn't seem like we're passing that information down in the converter call `.toTabs`. Up to you if that's easier to handle media outside of this.

[the remark this answers is not in the document]
Comment (boek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:39]: @jonalmeida I'm not sure I understand what you're saying. Tabs don't hold any mediastate so what would we pass into the converter?

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt (hunk 2 above)
Comment (boek) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:42]: @jonalmeida We're not taking any tabs off of browserState here. `this.tabs` was set with the `tabFilter`

File: components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt:39]: We spoke about this offline, but sharing here for completeness:

> You can also go about this 
>
> another way (in terms of a cleaner API to consume, arguably at least): Make another "collect" function that's invoked from the media scope, match the media aggregate data to the tabs, and then modify the `toTabs`<sup>[0]</sup> extension function to pass that media data into that data class. So when you get the `tabsTray.updateTabs`<sup>[1]</sup> call, you can update your ViewHolder state.
> 
> [0]: https://github.com/mozilla-mobile/android-components/blob/master/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/ext/BrowserState.kt#L11
> [1]: https://github.com/mozilla-mobile/android-components/blob/master/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt#L83

**Comment by jonalmeida** [comment]
bors retry

**Pull request merged**